lib/ex_doc/formatter/html/autolink.ex

```diff
@@ -311,171 +461,289 @@ defmodule ExDoc.Formatter.HTML.Autolink do
 
-  Create links to locally defined functions, specified in `locals`
-  as a list of `fun/arity` strings.
+  @regexes (for link_type <- @link_types,
+                language <- @languages,
+                kind <- @kinds do
+              %{
+                kind: kind,
+                language: language,
+                link_type: link_type
+              }
+            end)
 
-  Ignores functions which are already wrapped in markdown url syntax,
-  e.g. `[test/1](url)`. If the function doesn't touch the leading
-  or trailing `]`, e.g. `[my link link/1 is here](url)`, the fun/arity
-  will get translated to the new href of the function.
-  """
-  def locals(bin, locals, aliases \\ [], extension \\ ".html", lib_dirs \\ elixir_lib_dirs()) do
-    fun_re =
-      Regex.source(
-        ~r{(([ct]:)?([a-z_]+[A-Za-z_\d]*[\\?\\!]?|[\{\}=&\\|\\.<>~*^@\\+\\%\\!-\/]+)/\d+)}
-      )
+  @doc false
+  def locals(string, locals, aliases \\ [], extension \\ ".html", lib_dirs \\ elixir_lib_dirs()) do
+    options = [
+      locals: locals,
+      aliases: aliases,
+      extension: extension,
+      lib_dirs: lib_dirs
+    ]
 
-    regex = ~r{(?<!\[)`\s*(#{fun_re})\s*`(?!\])}
-    elixir_docs = get_elixir_docs(aliases, lib_dirs)
+    link(string, :elixir, :function, options)
+  end
 
-    Regex.replace(regex, bin, fn all, match ->
-      {prefix, _, function, arity} = split_function(match)
-      text = "`#{function}/#{arity}`"
+  @doc false
+  def elixir_modules(
+        string,
+        module_refs,
+        module_id \\ nil,
+        extension \\ ".html",
+        lib_dirs \\ elixir_lib_dirs()
+      )
+      when is_binary(string) do
+    options = [
+      module_refs: module_refs,
+      module_id: module_id,
+      extension: extension,
+      lib_dirs: lib_dirs
+    ]
 
-      cond do
-        match in locals ->
-          "[#{text}](##{prefix}#{enc_h(function)}/#{arity})"
+    link(string, :elixir, :module, options)
+  end
 
-        match in @basic_type_strings ->
-          "[#{text}](#{elixir_docs}#{@basic_types_page})"
+  @doc false
+  def elixir_functions(string, docs_refs, extension \\ ".html", lib_dirs \\ elixir_lib_dirs())
+      when is_binary(string) do
+    options = [
+      docs_refs: docs_refs,
+      extension: extension,
+      lib_dirs: lib_dirs
+    ]
 
-        match in @built_in_type_strings ->
-          "[#{text}](#{elixir_docs}#{@built_in_types_page})"
+    link(string, :elixir, :function, options)
+  end
 
-        match in @kernel_function_strings ->
-          "[#{text}](#{elixir_docs}Kernel#{extension}##{prefix}#{enc_h(function)}/#{arity})"
+  @doc false
+  def erlang_modules(string) when is_binary(string) do
+    link(string, :erlang, :module, [])
+  end
 
-        match in @special_form_strings ->
-          "[#{text}](#{elixir_docs}Kernel.SpecialForms" <>
-            "#{extension}##{prefix}#{enc_h(function)}/#{arity})"
+  @doc false
+  def erlang_functions(string) when is_binary(string) do
```

Comment:
I have made these set of functions no longer accesible in the documentation, as I consider they are not longer necessary. They have been kept because all the tests rely on them.
We should improve the docs of the newly introduce function that replaces them `link/4`.

Reply:
Those functions can be removed. We had them out to ease testing.